### PR TITLE
Add ConstrainedResetMap

### PR DESCRIPTION
**Pull request by schillic**

### CODE REVIEW

File: src/maps.jl [hunk 1]
```diff
@@ -325,3 +326,39 @@ function apply(m::ResetMap, x)
     end
     return y
 end
+
+"""
+    ConstrainedResetMap
+
+A reset mapwith state constraints of the form:
```
Comment (mforets) [inline src/maps.jl]:
```suggestion
A reset map with state constraints of the form:
```

File: src/maps.jl [hunk 2]
```diff
@@ -325,3 +326,39 @@ function apply(m::ResetMap, x)
...
+inputdim(::ConstrainedResetMap) = 0
+outputdim(m::ConstrainedResetMap) = m.dim
+islinear(::ConstrainedResetMap) = false
+isaffine(::ConstrainedResetMap) = true
+
+# convenience constructor for a list of pairs instead of a dictionary
+ConstrainedResetMap(dim::Int, X::ST, args::Pair{Int, <:N}...) where {N, ST} =
+    ConstrainedResetMap(dim, X, Dict{Int, N}(args))
+
+apply(m::ConstrainedResetMap, x) = apply(ResetMap(m.dim, m.dict), x)
```
Comment (mforets) [inline src/maps.jl]: While this is correct, it allocates a new reset map. How about using a union type in the existing  `apply` method for `ResetMap`?

Comment (schillic) [reply src/maps.jl]: Yes, I was not sure. Should I change this for all `ConstrainedX` types?

File: src/maps.jl [hunk 3]
```diff
@@ -325,3 +326,39 @@ function apply(m::ResetMap, x)
...
+variables are unchanged.
+
+### Fields
+
+- `dim`  -- dimension
+- `X`    -- state constraints
+- `dict` -- dictionary whose keys are the indices of the variables that are
+            reset, and whose values are the new values
+"""
+struct ConstrainedResetMap{N, ST} <: AbstractMap
```
Comment (mforets) [inline src/maps.jl]: This is just a whim, but can we keep the same order as for the `ResetMap`? `dim, dict, X` so

```julia
julia> m = ConstrainedResetMap(10, Dict(9 => 0.), X)
```

Comment (schillic) [reply src/maps.jl]: Then we cannot offer the convenience constructor.

File: src/maps.jl (hunk 2 above)
Comment (mforets) [reply src/maps.jl]: No, i think that's fine.. since they are one-liners ..

File: src/maps.jl (hunk 3 above)
Comment (mforets) [reply src/maps.jl]: Ah, ok i see. Then we forget my comment..

**Pull request merged**